examples/step-77/step-77.cc

```diff
@@ -0,0 +1,660 @@
+    // this function, rather than as a member variable of the surrounding class.
+    // We omit this step here because using the same coding style as in previous
+    // tutorial programs breeds familiarity with the common style and helps make
+    // these tutorial programs easier to read.)
+    {
+      TimerOutput::Scope t(computing_timer, "factorizing the Jacobian");
+
+      std::cout << "  Factorizing Jacobian matrix" << std::endl;
+
+      jacobian_matrix_factorization = std::make_unique<SparseDirectUMFPACK>();
+      jacobian_matrix_factorization->factorize(jacobian_matrix);
+    }
+  }
+
+
+
+  // @sect4{Computing the residual vector}
+
+  // The second part of what `assemble_system()` used to do in step-15 is
+  // computing the residual vector, i.e., the right hand side vector of the
+  // Newton linear systems. We have broken this out of the previous function,
+  // but the following function will be easy to understand if you understood
+  // what `assemble_system()` in step-15 did. Importantly, however, we need to
+  // compute the residual not linearized around the current solution vector, but
+  // whatever we get from KINSOL. This is necessary for operations such as line
+  // search where we want to know what the residual $F(U^k + \alpha_k \delta
+  // U^K)$ is for different values of $\alpha_k$; KINSOL in those cases simply
+  // gives us the argument to the function $F$ and we then compute the residual
+  // $F(\cdot)$ at this point.
+  //
+  // The function prints the norm of the so-computed residual at the end as a
+  // way for us to follow along the progress of the program.
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::compute_residual(
+    const Vector<double> &evaluation_point,
+    Vector<double> &      residual)
+  {
+    TimerOutput::Scope t(computing_timer, "assembling the residual");
+
+    std::cout << "  Computing residual vector..." << std::flush;
+
+    const QGauss<dim> quadrature_formula(fe.degree + 1);
+    FEValues<dim>     fe_values(fe,
+                            quadrature_formula,
+                            update_gradients | update_quadrature_points |
+                              update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe.n_dofs_per_cell();
+    const unsigned int n_q_points    = quadrature_formula.size();
+
+    Vector<double>              cell_residual(dofs_per_cell);
+    std::vector<Tensor<1, dim>> evaluation_point_gradients(n_q_points);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        cell_residual = 0;
+        fe_values.reinit(cell);
+
+        fe_values.get_function_gradients(evaluation_point,
+                                         evaluation_point_gradients);
+
+
+        for (unsigned int q = 0; q < n_q_points; ++q)
+          {
+            const double coeff =
+              1. / std::sqrt(1 + evaluation_point_gradients[q] *
```

Comment:
```suggestion
              1.0 / std::sqrt(1 + evaluation_point_gradients[q] *
```
This is a real nitpick, but line 243 has it written this way.